xds/src/main/java/io/grpc/xds/internal/certprovider/CertProviderSslContextProvider.java

```diff
@@ -0,0 +1,157 @@
+  @Nullable protected List<X509Certificate> savedTrustedRoots;
+
+  protected CertProviderSslContextProvider(
+      Node node,
+      Map<String, CertificateProviderInfo> certProviders,
+      CertificateProviderInstance certInstance,
+      CertificateProviderInstance rootCertInstance,
+      CertificateValidationContext staticCertValidationContext,
+      BaseTlsContext tlsContext,
+      CertificateProviderStore certificateProviderStore) {
+    super(tlsContext, staticCertValidationContext);
+    this.certInstance = certInstance;
+    this.rootCertInstance = rootCertInstance;
+    String certInstanceName = null;
+    if (certInstance != null && certInstance.isInitialized()) {
+      certInstanceName = certInstance.getInstanceName();
+      CertificateProviderInfo certProviderInstanceConfig =
+          getCertProviderConfig(certProviders, certInstanceName);
+      certHandle =
+          certificateProviderStore.createOrGetProvider(
+              certInstance.getCertificateName(),
+              certProviderInstanceConfig.getPluginName(),
+              certProviderInstanceConfig.getConfig(),
+              this,
+              true);
+    } else {
+      certHandle = null;
+    }
+    if (rootCertInstance != null
+        && rootCertInstance.isInitialized()
+        && !rootCertInstance.getInstanceName().equals(certInstanceName)) {
+      CertificateProviderInfo certProviderInstanceConfig =
+          getCertProviderConfig(certProviders, rootCertInstance.getInstanceName());
+      rootCertHandle =
+          certificateProviderStore.createOrGetProvider(
+              rootCertInstance.getCertificateName(),
+              certProviderInstanceConfig.getPluginName(),
+              certProviderInstanceConfig.getConfig(),
+              this,
+              true);
+    } else {
+      rootCertHandle = null;
+    }
+  }
+
+  private CertificateProviderInfo getCertProviderConfig(
+      Map<String, CertificateProviderInfo> certProviders, String pluginInstanceName) {
+    return certProviders.get(pluginInstanceName);
+  }
+
+  @Override
+  public final void updateCertificate(PrivateKey key, List<X509Certificate> certChain) {
+    savedKey = key;
+    savedCertChain = certChain;
+    updateSslContextWhenReady();
+  }
+
+  @Override
+  public final void updateTrustedRoots(List<X509Certificate> trustedRoots) {
+    savedTrustedRoots = trustedRoots;
+    updateSslContextWhenReady();
+  }
+
+  private void updateSslContextWhenReady() {
+    if (isMtls()) {
+      if (savedKey != null && savedTrustedRoots != null) {
+        updateSslContext();
+        clearKeysAndCerts();
+      }
+    } else if (isClientSideTls()) {
+      if (savedTrustedRoots != null) {
+        updateSslContext();
+        clearKeysAndCerts();
+      }
+    } else if (isServerSideTls()) {
+      if (savedKey != null) {
+        updateSslContext();
+        clearKeysAndCerts();
+      }
+    } else {
+      throw new IllegalStateException("Config is neither mTLS nor TLS!");
```

Comment:
Throwing such an exception here seem to be too late, and it's somewhat misleading. Callers of `updateCertificate()` and `updateTrustedRoots()` will never be the cause of this class being in invalid state. Instead, you should perform the check in the constructor.

Reply:
good point. fixed